### PR TITLE
[daal-2020-mnt] Add conformance tests in public ci

### PR DESCRIPTION
**Pull request by Pahandrovich**
This pull request contains automatic conformance testing with scikit-learn.

### CODE REVIEW

**Comment by napetrov** [comment]
I also assume that we have plans to get this in to master and releases?

File: .ci/conformance/tools/elastic_net/download_test.sh [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+echo
+echo "Test elastic net regression download started."
+echo
+wget -O test_elastic_net.py https://raw.githubusercontent.com/scikit-learn/scikit-learn/0.21.3/sklearn/linear_model/tests/test_coordinate_descent.py
```
Comment (napetrov) [inline .ci/conformance/tools/elastic_net/download_test.sh]: This is a lot of  duplication for this actions in multiple separate scripts. will be hard to made changes and align here.

File: .ci/conformance/tools/dbscan/download_test.sh [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+echo
+echo "Test dbscan download started."
+echo
+wget -O test_dbscan.py https://raw.githubusercontent.com/scikit-learn/scikit-learn/master/sklearn/cluster/tests/test_dbscan.py
```
Comment (napetrov) [inline .ci/conformance/tools/dbscan/download_test.sh]: This means that here you will use head of master instead of particular version as done in other places?

File: .ci/conformance/tools/elastic_net/patcher.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import re
+import sys
+
```
Comment (napetrov) [inline .ci/conformance/tools/elastic_net/patcher.py]: Is it really good idea to have multiple instances  of scripts for each alg? 

Also - why we make separate scripts here and not using daal4py mechanisms?

File: .ci/pipeline/ci.yml [hunk 4]
```diff
@@ -36,6 +36,72 @@ jobs:
     inputs:
       artifactName: '$(platform.type) build'
       targetPath: '__release_lnx_gnu'
+# conformance
```
Comment (napetrov) [inline .ci/pipeline/ci.yml]: This is too long

File: .ci/pipeline/ci.yml [hunk 5]
```diff
@@ -36,6 +36,72 @@ jobs:
...
+  - script: |
+      which git
+      git --version
+      wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh -O miniconda.sh;
```
Comment (napetrov) [inline .ci/pipeline/ci.yml]: This should be placed in scripts and not in pipeline - you can take a look on daal4py or scripts in master as examples

**Comment by Pahandrovich** [comment]
@napetrov Yes.

**Comment by napetrov** [comment]
make test

File: .ci/conformance/download_tests.sh [hunk 6]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+
+wget -O test_adaboost.py https://raw.githubusercontent.com/scikit-learn/scikit-learn/0.21.3/sklearn/ensemble/tests/test_weight_boosting.py
```
Comment (napetrov) [inline .ci/conformance/download_tests.sh]: suggest to make this a variable - "https://raw.githubusercontent.com/scikit-learn/scikit-learn"

make this as another variable - 0.21.3

File: .ci/conformance/run_tests.sh [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+echo "Testing started `date +'%d.%m.%y %H:%M:%S'`."
+echo
+
+file_algs=algorithms.txt
+
+while read ALGORITHM_NAME
+    do 
+        echo Testing $ALGORITHM_NAME
+        echo
+        source ./tools/${ALGORITHM_NAME}/download_test.sh &>> _log_downloads
```
Comment (napetrov) [inline .ci/conformance/run_tests.sh]: Think common download script should be used now? 
ci/conformance/download_tests.sh

File: .ci/conformance/tools/dbscan/patcher.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+                break
+
+    f = open(file_name, "w")
+    f.writelines(lines)
+    f.close()
+
+
+if __name__ == "__main__":
+    path_file(fil_to_path, dictionary)
+
```
Comment (napetrov) [inline .ci/conformance/tools/dbscan/patcher.py]: If you would like to keep separate patch files, might be at least move common part to centralized module? 

Or alternatively you can make single patcher script that will accept alg as param.

File: .ci/conformance/tools/elastic_net/patcher.py (hunk 3 above)
Comment (Pahandrovich) [reply .ci/conformance/tools/elastic_net/patcher.py]: We use wrappers from daal4py, but modify them by inserting call counts.

File: .ci/conformance/wrappers/wrapper_lin_reg.py [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+        return "float"
+    else:
+        raise ValueError("Input array has unexpected dtype = {}".format(dt))
+
+def make2d(X):
+    if np.isscalar(X):
+        X = np.asarray(X)[np.newaxis, np.newaxis]
+    elif isinstance(X, np.ndarray) and X.ndim == 1:
+        X = X.reshape((X.size, 1))
+    return X
```
Comment (prutskov) [inline .ci/conformance/wrappers/wrapper_lin_reg.py]: this functions already exists in`.ci/conformance/wrappers/utils.py`

File: .ci/conformance/wrappers/wrapper_svm.py [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+        if self.decision_function_shape is 'ovo':
+            df_shape_ovo_calls += 1
+
+        if self.decision_function_shape is 'ovr':
+            df_shape_ovr_calls += 1
+
+        if self.random_state is not None:
+            random_state_calls += 1
+
+        print_counters()
```
Comment (prutskov) [inline .ci/conformance/wrappers/wrapper_svm.py]: I think that function like this applicable to use in all wrappers.

**Comment by Pahandrovich** [comment]
@prutskov @napetrov @PivovarA Any other concern?

File: .ci/pipeline/ci.yml [hunk 11]
```diff
@@ -36,6 +36,14 @@ jobs:
     inputs:
       artifactName: '$(platform.type) build'
       targetPath: '__release_lnx_gnu'
+  - script: |
+      .ci/scripts/conformance_test.sh --build-dir __release_lnx_gnu
```
Comment (napetrov) [inline .ci/pipeline/ci.yml]: In this case conformance will be executed for both 32 and 64 bits. Does 32 bits make sense?

File: .ci/conformance/tools/elastic_net/patcher.py (hunk 3 above)
Comment (napetrov) [reply .ci/conformance/tools/elastic_net/patcher.py]: Might be it worth to place this code in to daal4py wrappers instead of inserting call counts on top of it? 
for example conformance mode can be enabled in daal4py wrappers as non defalt

File: .ci/conformance/wrappers/logistic_loss.py [hunk 12]
```diff
@@ -0,0 +1,167 @@
+#
+#*******************************************************************************
```
[the remark this answers is not in the document]
Comment (napetrov) [reply .ci/conformance/wrappers/logistic_loss.py]: I don't sure that i get question, but hope that @Pahandrovich is going to answer it?

**Comment by napetrov** [comment]
> 
> 
> @prutskov @napetrov @PivovarA Any other concern?

I think not all comments from @prutskov  have been adressed

File: .ci/conformance/wrappers/logistic_loss.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (prutskov) [reply .ci/conformance/wrappers/logistic_loss.py:2]: Will we have any legal problems if we have code from daal4py/ license header in some files? As I remember this file is full copy of file from daal4py repo.

File: .ci/conformance/tools/elastic_net/patcher.py (hunk 3 above)
Comment (Pahandrovich) [reply .ci/conformance/tools/elastic_net/patcher.py]: @napetrov Do you mean to add counters to the daal4py repository itself?

File: .ci/conformance/wrappers/logistic_loss.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Pahandrovich) [reply .ci/conformance/wrappers/logistic_loss.py:2]: I think I can take this file from the daal4py folder.

**Comment by napetrov** [comment]
> 
> 
> @prutskov @napetrov @PivovarA Any other concern?

I think not all comments from @prutskov  have been addressed

File: .ci/conformance/tools/elastic_net/patcher.py (hunk 3 above)
Comment (napetrov) [reply .ci/conformance/tools/elastic_net/patcher.py]: yes
